File: dev/armadillobenchmark/src/01_matrix_calculation.cpp
#include "00_main.h"

// This was adapted from
// https://github.com/rbenchmark/benchmarks/blob/master/R-benchmark-25/
// The changes are:
// 1. Rewriting the tests in Armadillo allowing to change the matrix size
// 2. Moving the Armadillo tests to base R in a comparable way (i.e.,
//    using qr.solve() instead of lm())

// Creation, transposition, deformation of an NxN matrix

[[cpp11::register]] int matrix_calculation_01_arma_(const int& n) {
  mat a = randn<mat>(n, n) / 10;
  mat b = a.t();
  b.reshape(n / 2, n * 2);
  return 0;
}

// NxN normal distributed random matrix^1000

[[cpp11::register]] int matrix_calculation_02_arma_(const int& n) {
  mat a = abs(randn<mat>(n, n) / 2);
  mat b = pow(a, 1000);
  return 0;
}

// Sorting of N random values

[[cpp11::register]] int matrix_calculation_03_arma_(const int& n) {
  vec a = randn<vec>(n);
  a = sort(a);
  return 0;
}

// NxN cross-product matrix

[[cpp11::register]] int matrix_calculation_04_arma_(const int& n) {
  mat a = randn<mat>(n, n);
  mat b = a.t() * a;
  return 0;
}

// Linear regression over an NxN matrix

[[cpp11::register]] int matrix_calculation_05_arma_(const int& n) {
  mat x = randn<mat>(n, n);
  vec y = randn<mat>(n);
  vec c = solve(x.t(), x.t() * y);
  return 0;
}

File: dev/basebenchmark/R/01-matrix-calculation.R
# Creation, transposition, deformation of an NxN matrix ----

#' @export 
matrix_calculation_01_r <- function(n) {
  a <- matrix(rnorm(n * n) / 10, ncol = n, nrow = n)
  b <- t(a)
  dim(b) <- c(n/2, n*2)
  a <- t(b)
  return(0L)
}

# NxN normal distributed random matrix^1000 ----

#' @export
matrix_calculation_02_r <- function(n) {
  a <- abs(matrix(rnorm(n * n) / 2, ncol = n, nrow = n))
  b <- a^1000
  return(0L)
}

# Sorting of N random values ----

#' @export
matrix_calculation_03_r <- function(n) {
  a <- rnorm(n)
  b <- sort(a, method = "quick")
  return(0L)
}

# NxN cross-product matrix ----

#' @export
matrix_calculation_04_r <- function(n) {
  a <- matrix(rnorm(n * n), ncol = n, nrow = n)
  b <- crossprod(a)
  return(0L)
}

# Linear regression over an NxN matrix via QR decomposition ----

#' @export
matrix_calculation_05_r <- function(n) {
  # same as lm() for full rank matrices but in Armadillo there is no lm()
  # equivalent
  x <- matrix(rnorm(n * n), ncol = n, nrow = n)
  y <- rnorm(n)
  b <- qr.solve(x, y)
  return(0L)
}
